rfd/0163-vnet.md

```diff
@@ -0,0 +1,695 @@
+
+If VNet is not running, starting VNet from the VNet tab starts it only for the root cluster of the
+current workspace. Likewise, stopping it from the VNet tab stops it only for the current root
+cluster and turns off autostart just for this cluster.
+
+This behavior aligns with a couple of ideas:
+
+1) The connection list is a UI element that is not scoped to any workspace, unlike tabs which are
+always scoped to a specific workspace.
+2) The user can turn off VNet for a specific workspace if they do not wish to use it in that
+workspace.
+
+##### Expired certs
+
+As shown in the diagram in the [DNS queries and IP address
+assignment](#dns-queries-and-ip-address-assignment) section, there are two calls made to the
+cluster, one to get the details of an app and one to proxy a connection to the cluster. If any of
+those calls fails due to expired certs, Connect prompts the user to log in again. It's similar to
+how Connect prompts for relogin after the cert of a local db proxy expires.
+
+##### Considered alternatives
+
+###### Launch VNet on system start
+
+Doing so would require some kind of persistent UI that is also launched on system start, perhaps an
+app in the Menu Bar. The UI would need to enable the user to manage the state of VNet and provide a
+way to refresh expired certs. Perhaps Connect could support headless mode where it does not open the
+main window automatically and instead launches a separate frontend app which handles relogin. Alas,
+we don't have a Menu Bar app now and Connect is not prepared to launch in headless mode at the
+moment.
+
+###### Ask for root privileges just once
+
+Any of the alternatives described in the [Alternatives to osascript](#alternatives-to-osascript)
+section would allow us to ask for root privileges just once at the cost of development time.
+
+#### tsh integration
+
+`tsh vnet` prompts for a password and sets up VNet for the current profile. The `--all` (`-R`) flag
+can be used to set up VNet for all non-expired profiles. Only one instance of VNet can be active
+at a time, guarded by a lockfile in the tsh directory. To allow multiple VNet instances running at
+the same time, tsh processes would need to coordinate which TUN device to use and how to set up DNS.
+
+The initial output includes a list of proxied DNS zones. The format of the list depends on the
+number of profiles VNet has been configured for and the number of leaf clusters.
+
+<details>
+<summary>Different tsh vnet output formats</summary>
+
+One cluster with no leaf clusters:
+
+```
+$ tsh vnet
+
+Proxying connections made to .teleport-local.dev.internal, .company.private.
+```
+
+One cluster with leafs:
+
+```
+$ tsh vnet
+
+Proxying connections made to the DNS zones listed below.
+
+DNS Zone                         Cluster
+-------------------------------- ----------------------
+.teleport.example.com.internal   teleport.example.com
+.example.com                     teleport.example.com
```

Comment:
I'm not sure if the nomenclature here is correct and if this is the format with which we want to represent DNS zones.